plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/repository/AbstractSourceAttachmentDownloaderJob.java

```diff
@@ -0,0 +1,142 @@
+package com.google.cloud.tools.eclipse.appengine.libraries.repository;
+
+import com.google.cloud.tools.eclipse.appengine.libraries.LibraryClasspathContainer;
+import com.google.cloud.tools.eclipse.appengine.libraries.persistence.LibraryClasspathContainerSerializer;
+import com.google.cloud.tools.eclipse.util.status.StatusUtil;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import javax.annotation.PostConstruct;
+import javax.inject.Inject;
+import javax.inject.Named;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.jobs.Job;
+import org.eclipse.jdt.core.IClasspathContainer;
+import org.eclipse.jdt.core.IClasspathEntry;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.core.JavaModelException;
+
+public abstract class AbstractSourceAttachmentDownloaderJob extends Job {
+
+  private long retryDelayFactor = 3000L;
```

Comment:
It seems that the problem of getting an unresolved container while trying to attach source path was caused by the `@PostConstruct` method of the source downloader jobs not being called and thus exclusive access to the project was not enforced. The fact that the `@PostConstruct` method was not called is caused by a known issue where the @PostConstruct class loaded for the DI processing mechanism and the class loaded for the job instance were considered different because (probably) they were loaded by different class loaders. This is a [known issue](http://wiki.eclipse.org/Eclipse4/RCP/FAQ#Why_aren.27t_my_.40Inject-able.2F.40PostConstruct_methods_being_injected.3F) which also interacts with another problem when [`guava` is on the classpath](https://www.eclipse.org/forums/index.php/t/1071393/) .
I find the @PostConstruct invocation flaky/non-deterministic, so I'll change the job creation to make sure that `setRule()` is called to ensure exclusive access to the project being modified by the resolver/source downloader jobs.